### DISCUSSION TITLE
"Reserve as you use"

### DISCUSSION THREAD

**Discussion by florent-uzio** [XRPLF/XRPL-Standards#126] [XLS Ideas (pre standard proposal)]
The ongoing discussion revolves around the XRP reserve and the entry cost for the XRP Ledger, particularly in the context of the current value of 1 XRP at around $0.6281 USD as of August 14, 2023.

**Engineering Perspective** 

Ensuring the stability of the XRPL network remains a top priority. This requires the servers running rippled to possess sufficient hardware capacity, encompassing CPU, RAM, bandwidth, storage, and more. 

The implementation of reserves serves as an anti-spam mechanism, preventing excessive creation of accounts that could overload servers and the XRPL network. Maintaining an equitable entry point through reserves remains essential. 

**Business Perspective** 

This aspect is particularly pertinent to firms that safeguard customer private seeds, aiming to offer user-friendly applications and seamless experiences.

Feedback from these companies indicates that serving customers via XRPL becomes progressively expensive as the number of XRPL accounts increases. 

For instance, targeting 10,000 customers necessitates an allocation of approximately $62,810 USD at the current rate, a notable sum even for a relatively modest user base.

Furthermore, inactive accounts (with only a couple of transactions done) can pose a challenge as the cost of opening such accounts remains substantial (10 XRP).

**Proposed Solution: "Reserve as you use"** 

Introducing the concept of "Reserve as you use" seeks to address these concerns. The idea is straightforward: the more a user engages with their XRPL account (conducting transactions), the more XRP is assigned to the base reserve.

This approach would result in a lower entry base reserve while the allocation to the base reserve adheres to a function (e.g., f(x)) with an upper limit set at the current XRP reserve (10 XRP).

**Potential Functions** 

Several functions can be considered, such as `ln(x+2)`, `ln((x^2) + 2)`, or `ln((x^3) + 2)`. The chosen function would determine the allocation of XRP to the base reserve based on the number of transactions conducted by the user. 

**Example: Using ln((x^2) + 2)** 

The following table showcases the XRP allocation to the base reserve based on the number of transactions (`x`) submitted to the XRPL using the function `ln((x^2) + 2)`: 

| x - Number of XRPL Transactions | ln((x^2) + 2) - XRP allocated to base reserve | 
|---------------------:|:----------------------:| 
|   0 | 0.6931471806  |
|   1 | 1.098612289   |
|   2 | 1.386294361   |
|   5 | 3.218875825   |
|  10 | 4.605170186   |
|  30 | 6.802394763   |
|  50 | 7.824046011   |
| 100 | 9.210340372   |
| 150 | 10.02127059 (capped at 10 XRP)   |
| 200 | 10.59663473 (capped at 10 XRP)   |

![ln(x^2 + 2)](https://github.com/XRPLF/XRPL-Standards/assets/36513774/8383b317-d4fb-4732-be5f-18e9baef8e59)

In this scenario, employing the `ln((x^2) + 2)` function sets the entry price at ~0.70 XRP. Consequently, a user can perform nearly 150 transactions to reach the 10 XRP base reserve. Subsequent transactions would then incur standard fees per transaction (0.000012 XRP per payment for example). 

**Conclusion** 

Implementing the "Reserve as you use" approach would enable a more cost-effective entry for companies into the XRPL while adapting the base reserve allocation according to account activity. 

This proposal attempts to address both engineering stability and business cost concerns, contributing to the XRP Ledger's attractiveness for a growing range of companies.

**Comment by WietseWind** [comment]
While I completely agree at the threshold to gain adoption/users with the current reserve model, the proposed model (thanks for kicking off this conversation!) does incentivize creating multiple accounts without sufficiently disincentivizing abandoning the account. It would make "throwaway accounts" much cheaper, resulting in ledger bloat and storage costs for infra contributors.

**Reply by florent-uzio** [reply]
Thanks Wietse, the numbers/functions above are just examples and can be changed.
The entry cost to enable an account could be adapted and set to a higher XRP value to prevent people from creating throwaway accounts while keeping it low enough to make it attractive.

**Reply by WietseWind** [reply]
I understand. The challenge is that the sweet spot between nog interesting enough for throwaway account, vs. still viable for onboarding is hard. Even harder if you add the global scale of a blockchain into the mix, where e.g. 4 XRP in Western Europe / US is OK, but e.g. 4 XRP in some parts of the world (where the benefits may be highest, more unbanked people, etc.) is relatively high.

It would be great (but this topic has been revisited often) if a drastic change like dynamic reserves would yield a viable solution not only in the rich(er) parts of the world.

I don't have the answer, so it's great to get this discussion started again, but I feel we should look at this discussion with not only ledger/infra const / incentives in mind, but also inclusivity on a larger (global) scale.

**Reply by mvadari** [reply]
Just thinking "out loud" here: what if this kind of solution was coupled with reducing the cost of AccountDelete? Let's say it costs 2 XRP (the object reserve) for an account with no transactions, but you get 1.5 XRP back (and only lose 0.5 XRP) by deleting your account, thereby incentivizing people to delete their throwaway accounts instead of leaving them around as ledger bloat.
This would then only be a FH issue, but those transactions would likely have occurred anyways, just with the user's main account instead of their throwaway account.

**Reply by florent-uzio** [reply]
That's a valid point regarding inclusivity.

Just checking, can an account be deleted only if it doesn't have transactions? I am trying to delete [this account](https://testnet.xrpl.org/accounts/rnwrLRv2sZBNqKHQN9P4KQLMPVLoXeNYvD) but it fails. On the other hand an account freshly created can be deleted. Update: after trying again after a couple of hours, the deletion worked.

The idea of reducing the cost of AccountDelete is interesting with a minimum base entry at 2 XRP.

Do we have any data/metrics about the number of dormant/throwaway accounts on mainnet?

Throwing another idea here, I feel that defining the reserves in XRP will always put us in a situation where we need to adjust it with the XRP price fluctuation.
How hard would it be to rethink the whole reserve logic by basing it on the USD price, or at least a USD range price? rippled would automatically adjust the reserve based on the price from multiple sources. Could we leverage the future Oracle feature for this?

The numbers are completely made up but for example:

| XRP Price Range (USD) | Base Reserve (XRP) | Owner Reserve (XRP) | 
|-----------------------|--------------|---------------|
| 0-0.2                 | 10           | 2             |
| 0.2-0.5               | 8            | 1.5           |
| 0.5-1                 | 4            | 1             |

**Reply by mvadari** [reply]
I really don't like the idea of pinning it to the USD price, mostly because this is a core feature of a decentralized blockchain and there isn't really a decentralized trustworthy way to check the current price of XRP with respect to USD (I guess you could use the DEX, but it would be very difficult to prevent that from being gameable).
Isn't that kind of the point of having these values be adjustable? The UNL can adjust the reserve based on the current price of XRP.

**Reply by florent-uzio** [reply]
I understand.
When you mention
> Isn't that kind of the point of having these values be adjustable? The UNL can adjust the reserve based on the current price of XRP.

The fact that those reserve values are adjustable is great but it is a manual process at the moment to update them, isn't it?
The last change from 20 to 10 XRP was done approximately 2 years ago right?
If the process to update the reserve is manual, would it be possible to make it automatic like every 3 months?

**Reply by SoulSeige** [reply]
Just a thought on reducing dormant accounts:

What if all accounts that are under the current reserve are deleted when they go dormant.

For example, with "Reserve as you go", unless the users reserve hits the "cap" (10 XRP), their account will be deleted if over 30 days (or lets say X amount of ledgers) pass without the account posting a transaction.

**Reply by mvadari** [reply]
What do you do with the user's leftover XRP, then? You can't burn it. How do you differentiate between dormant accounts and savings accounts (people just storing their XRP long-term)?

**Reply by SoulSeige** [reply]
People storing their XRP (Savings accounts) will have more than the reserve amount, so this wouldn't apply to them.

Interesting point on the leftover XRP. Why is it not possible to burn? The AccountDelete transaction currently burns 2 XRP (owner reserve), this could be treated in a similar manner?

If the concern is around excessive supply reduction, some other potential ideas could be:
-Allow users to set a "forwarding" address during AccountSet
-A potential future staking mechanism reward?
-A reward for the winner of the auction of the top traded AMM pool?
I realise these are fairly complex, and intertwine this idea with other amendments. Although I welcome more thoughts/ideas for this, as I think the dormant Account deletion acts as a great counterbalance to the reduced friction of creating an Account.

**Reply by mvadari** [reply]
The bigger issue is likely that nothing happens on the XRPL by itself, everything needs a trigger (even with hooks). So you can't just automatically delete accounts, someone has to send some sort of transaction that deletes them - just like how you need to send an EscrowFinish/EscrowCancel tx in order to transfer an escrow's funds. I guess in this case anyone could submit an AccountDelete tx to clean it up, but that's not the greatest implementation strategy.

[a comment by an automated account is not shown]
